[IMP] base: stop logging INFO access errors

kmagusiak opened the pull request:
There is no need to log an INFO every time we create an AccessError. The exception will be raised and the caller is responsible of either letting it propage or catching and handling it: by replacing the exception, logging, etc.

Current behavior before PR:
If you make an access error and catch it, you will see it in the logs. It is an INFO and there is no stacktrace, so it gives no information when accessing a related model from somewhere else.

Desired behavior after PR is merged:
Stop logging these messages.  They will be logged in the exception is not caught.


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/185753.png)](https://mergebot.odoo.com/odoo/odoo/pull/185753)

ryv-odoo commented:
Hi @odoo/rd-security, are you ok with these change ?

Julien00859 commented on odoo/http.py:
```diff
@@ -2435,10 +2435,12 @@ def fake_start_response(status, headers):
                 elif isinstance(exc, SessionExpiredException):
                     _logger.info(exc)
-                elif isinstance(exc, (UserError, AccessError)):
+                elif isinstance(exc, AccessError):
+                    _logger.warning(exc, exc_info=exc)
```
Wearing my @odoo/rd-http hat
Great! Love it!

Julien00859 commented:
```diff
diff --git a/addons/rpc/controllers.py b/addons/rpc/controllers.py
index d78ab59b5b13..0bf2a1ffa06f 100644
--- a/addons/rpc/controllers.py
+++ b/addons/rpc/controllers.py
@@ -149,6 +149,8 @@ class RPC(Controller):
         try:
             response = self._xmlrpc(service)
         except Exception as error:  # noqa: BLE001
+            if isinstance(error, odoo.exceptions.AccessError):
+                _logger.warning(...)
             response = xmlrpc_handle_exception_string(error)
         return Response(response=response, mimetype='text/xml')
 
@@ -159,6 +161,8 @@ class RPC(Controller):
         try:
             response = self._xmlrpc(service)
         except Exception as error:  # noqa: BLE001
+            if isinstance(error, odoo.exceptions.AccessError):
+                _logger.warning(...)
             response = xmlrpc_handle_exception_int(error)
         return Response(response=response, mimetype='text/xml')
```

kmagusiak commented:
@Julien00859 @xmo-odoo Currenly, if you execute a remote call using jsonrpc, you will get a stacktrace on the server, but if you make the same call using xmlrpc, you will not have a stacktrace. I have added a commit to align the way jsonrpc and xmlrpc are working so that when either one raises an exception it goes through the patch in http.py that handles exceptions (where some of them are logged). This avoids duplicating logging logic.

kmagusiak commented on addons/rpc/controllers.py:
```diff
@@ -150,6 +150,8 @@ def xmlrpc_1(self, service):
             response = xmlrpc_handle_exception_string(error)
+            error.error_response = Response(response=response, mimetype='text/xml')
+            raise
         return Response(response=response, mimetype='text/xml')
```
I have isolated that change in #193208. Because it needs fixing some tests and is not really related.

kmagusiak commented:
@beledouxdenis could you check this PR?

beledouxdenis commented:
In the pull request description, you do not mention clearly that tracebacks are now being added to the server log for every access error.
<details>
<summary>
I tried with both xmlrpc and json rpc, before and after the PR. Whether it's jsonrpc or xmlrpc, before there was no traceback along the access error in the server logs, and now there is.
</summary>

```py
#!/usr/bin/env python3
import requests
import xmlrpc.client

db, url = "master", "http://localhost:8069"
username, password = "portal", "portal"

session = requests.session()
result = session.post(f"{url}/web/session/authenticate", json={
    "params": {
        "db": db,
        "login": username,
        "password": password,
    }}).json()['result']
user_id = result.get('uid')
partner_id = result.get('partner_id')

user_id = xmlrpc.client.ServerProxy(f"{url}/xmlrpc/2/common").authenticate(db, username, password, {})
models = xmlrpc.client.ServerProxy(f"{url}/xmlrpc/2/object")


def xmlrpc(model, method, *args, **kwargs):
    return models.execute_kw(db, user_id, password, model, method, args, kwargs)


def jsonrpc(model, method, *args, **kwargs):
    response = session.post(f"{url}/web/dataset/call_kw", json={"params": {
        "model": model,
        "method": method,
        "args": args,
        "kwargs": kwargs,
    }})
    try:
        response = response.json()
    except Exception:
        print(response.text)
        raise Exception
    if 'error' in response:
        if response['error']['data']['name'] == 'odoo.exceptions.UserError':
            return response['error']['data']['arguments'][0]
        else:
            print(response['error']['data']['debug'])
            raise Exception
    return response["result"]


result = xmlrpc('res.users', 'read', [1], ["password"], context={})
print(result)

session.close()

```

</details>

I do not say this is wrong, it can be pretty handy for debugging or forensics purposes, but I think:
- It still worth mentioning in the PR description more clearly, this is a pretty big change, and this deserves a thorough thinking
- I wonder if it will not bloat the server logs. Access error are occuring pretty often and this might increase the server logs size (I mean in kilobytes) quite a bit, on odoo.com but on the saas as well, while the information will be worthless 99.9% of the time. That being said don't get me wrong, in case of actual attacks, on the contrary, the information could definitely be life saving. So maybe it's worth it to include it even if its worth less 99% of the time.
-  Maybe you should double check with the sentry team if it will not bother them / it will be out of their filters. There is a guideline that normal, or "expected", errors should not have a traceback, and maybe this falls into it.

In addition I join the Xavier's comment
> if there are other cases of access errors being logged as info and never warned and whether they are legitimate suppressions

I hope there are no other cases, where we will loose the access error log completely. I do not think this should be under-estimated 😬

kmagusiak commented:
> In the pull request description, you do not mention clearly that tracebacks are now being added to the server log for every access error.

Added.

> I tried with both xmlrpc and json rpc, before and after the PR. Whether it's jsonrpc or xmlrpc, before there was no traceback along the access error in the server logs, and now there is.

We fixed separately xmlrpc not re-raising exceptions to the WSGI app (`odoo.http`).

> I do not say this is wrong, it can be pretty handy for debugging or forensics purposes, but I think:
> 
> * It still worth mentioning in the PR description more clearly, this is a pretty big change, and this deserves a thorough thinking
> * I wonder if it will not bloat the server logs. Access error are occuring pretty often and this might increase the server logs size (I mean in kilobytes) quite a bit, on odoo.com but on the saas as well, while the information will be worthless 99.9% of the time. That being said don't get me wrong, in case of actual attacks, on the contrary, the information could definitely be life saving. So maybe it's worth it to include it even if its worth less 99% of the time.

I had a lot of cases at the client when someone said they did not have access and we were just blind during the investigation because we did not know where the call was coming from. Typically, no access to some part of stock or tasks when validating a sale order.

> * Maybe you should double check with the sentry team if it will not bother them / it will be out of their filters. There is a guideline that normal, or "expected", errors should not have a traceback, and maybe this falls into it.

@odoo/sentry_reviewers Could you give some input?

> In addition I join the Xavier's comment
> 
> > if there are other cases of access errors being logged as info and never warned and whether they are legitimate suppressions
> 
> I hope there are no other cases, where we will loose the access error log completely. I do not think this should be under-estimated 😬

I have checked the `except AccessError` blocks and they are checking whether the user has access (and returning a bool or equivalent), they log the error themselves (before doing something else) or they re-raise it. The cases I an not sure about is when in the except block there is a `raise ...NotFound()` or a redirect.

Julien00859 commented:
> The cases I an not sure about is when in the except block there is a raise ...NotFound() or a redirect.

For those ones it means that we are hiding the actual access error and pretend that the URL doesn't exist. We should evaluate case-by-case if we deem necessary to log something.

kmagusiak commented:
From sentry team:
> Both access errors and warnings are ignored, we don't sevent any events about those to Sentry
So it looks ok
> Thx for asking 🙏

@beledouxdenis ok to merge?

kmagusiak commented:
@odoo/rd-security Could you approve?

beledouxdenis commented:
I honestly cannot pronounce myself, the ratio risk / added value does not look good to me, the risk looks greater than the added value, and I currently do not have the resources to investigate / dig more if there is a great risk or not. In the end at worse we just loose logs, so its not that bad, its not such a high risk, but on the other hand logs are important for forensics and we will not know we are missing these logs before we actually need them, after an actual attack / malicious act. While the added value is like "just" code cleaning. Which is important but more important than risking loosing logs ?

So I am reserved, though I would tend to say yes, I would just feel re-assured if it was more battle-tested.

kmagusiak commented:
> I honestly cannot pronounce myself, the ratio risk / added value does not look good to me, the risk looks greater than the added value, and I currently do not have the resources to investigate / dig more if there is a great risk or not. In the end at worse we just loose logs, so its not that bad, its not such a high risk, but on the other hand logs are important for forensics and we will not know we are missing these logs before we actually need them, after an actual attack / malicious act. While the added value is like "just" code cleaning. Which is important but more important than risking loosing logs ?
> 
> So I am reserved, though I would tend to say yes, I would just feel re-assured if it was more battle-tested.

We raise an exception that will get logged unless some code caught and handled/ignored it. There is no stack trace, so you can get this on not directly related models and no idea where it comes from. 

It is basically the equivalent of https://stackoverflow.com/questions/6639963/why-is-log-and-throw-considered-an-anti-pattern
